File: algo/leetcode_155.cxx
#include <iostream>
#include <list>

using namespace std;

class MinStack {
private:
    list<long> elems;
    long min_val;
public: 
    void push(int x) {
        if (elems.empty()) {
            elems.push_front(0);            
            min_val = x;
        } else {
            long e = x - min_val;
            elems.push_front(e);
            if (e < 0) min_val = x;
        }
    }

    void pop() {
        long e = elems.front();
        if (e < 0) min_val -= e;
        elems.pop_front();
    }

    int top() {
        long e = elems.front();
        if (e > 0) {
            return (int)(e + min_val);
        } else {
            return (int)(min_val);
        }
    }
    
    int getMin() {
        return (int)(min_val);
    }
};

class MinStackDuo {
private:
    list<int> min_elems;
    list<int> ord_elems;

public:    
    void push(int x) {
        ord_elems.push_front(x);
        if (min_elems.empty() || 
            min_elems.front() >= x)
            min_elems.push_front(x);
    }
    
    void pop() {
        if (ord_elems.front() == min_elems.front())
            min_elems.pop_front();
        ord_elems.pop_front();
    }
    
    int top() {
        return ord_elems.front();
    }
    
    int getMin() {
        return min_elems.front();
    }
};


int main() {
    MinStack minStack;
    minStack.push(2);
    minStack.push(0);
    minStack.push(3);
    minStack.push(0);
    cout << minStack.getMin() << endl;   
    minStack.pop();
    cout << minStack.getMin() << endl;   
    minStack.pop();
    cout << minStack.getMin() << endl;   
    minStack.pop();
    cout << minStack.getMin() << endl;   
}

File: algo/leetcode_155.scala
class MinStack() {
  import scala.collection.mutable
  import java.util.EmptyStackException

  val minElems = mutable.ListBuffer.empty[Int]
  val ordElems = mutable.ListBuffer.empty[Int]

  def push(x: Int) = {
    ordElems += x
    if (minElems.isEmpty) {
      minElems += x
    } else if (minElems.last >= x) {
      minElems += x
    }
  }

  def pop() = {
    require(minElems.nonEmpty && ordElems.nonEmpty)
    if (minElems.last == ordElems.last)
      minElems.trimEnd(1)
    ordElems.trimEnd(1)
  }

  def top() : Int = {
    ordElems.lastOption.getOrElse {
      throw new EmptyStackException()
    }
  }

  def getMin() : Int = {
    minElems.lastOption.getOrElse {
      throw new EmptyStackException()
    }
  }
}

val minStack = new MinStack()
minStack.push(-2)
minStack.push(0)
minStack.push(-3)
println(minStack.getMin())
minStack.pop()
println(minStack.top())
println(minStack.getMin())
